### ISSUE TITLE
Problem inferring nested generic classes with defaults

### ISSUE THREAD

**Issue by cmp0xff** [microsoft/pyright#11187]
## Describe the bug
`pyright` cannot infer nested generic classes with default values properly.

## Code or Screenshots

```python
from typing import Any, Generic
from collections.abc import Sequence
from typing_extensions import TypeVar, Self, reveal_type

_OrderableT = TypeVar("_OrderableT", default=Any)

class Interval(Generic[_OrderableT]):
    def __new__(cls, left: _OrderableT, right: _OrderableT) -> Self: ...

IntervalT = TypeVar("IntervalT", bound=Interval, default=Interval)

S2 = TypeVar("S2")

class IntervalIndex(Generic[S2]):
    def __new__(cls, data: Sequence[IntervalT]) -> IntervalIndex[IntervalT]: ...

reveal_type(Interval(0, 1))  # type checkers all give Interval[int]
reveal_type(IntervalIndex([Interval(0, 1)]))  # mypy gives IntervalIndex[Interval[int]], pyright gives IntervalIndex[Interval[Any]]
```

## VS Code extension or command-line

Command line, `pyright==1.1.407`

**Comment by just-maiyak** [comment]
Hello !

I think I have a concrete example of this bug in the library `psycopg3`:

```py
import psycopg
from psycopg.rows import dict_row

def test():
    with psycopg.connect(
        row_factory=dict_row # This fails to typecheck
    ) as conn:
        pass
```

I have reproduced a minimal setup for this (mimicking what is done in the library):

```py
from typing import Any, Generic, TypeAlias, TypeVar, Protocol


TupleRow: TypeAlias = tuple[Any, ...]
DictRow: TypeAlias = dict[str, Any]
Row = TypeVar("Row", covariant=True, default="TupleRow")


class RowMaker(Protocol[Row]):
    def __call__(self, __test: Any) -> Row: ...


class RowFactory(Protocol[Row]):
    def __call__(self, __test: Any) -> RowMaker[Row]: ...


class Connection(Generic[Row]):
    @classmethod
    def connect(
        cls,
        conninfo: str = "",
        *,
        row_func: RowFactory[Row] | None = None
    ) -> None:
        pass


def dict_row(__x: Any) -> RowMaker[DictRow]: ...

Connection.connect(row_func=dict_row) # Does not typecheck under pyright but does with mypy
```

I get the following result upon inspection with pyright :

```
/Users/just.maiyak/Projects/rnd/submarine/test_pyright.py
  /Users/just.maiyak/Projects/rnd/submarine/test_pyright.py:30:29 - error: Argument of type "(__x: Any, /) -> RowMaker[DictRow]" cannot be assigned to parameter "row_func" of type "RowFactory[TupleRow] | None" in function "connect"
    Type "(__x: Any, /) -> RowMaker[DictRow]" is not assignable to type "RowFactory[TupleRow] | None"
      Type "(__x: Any, /) -> RowMaker[DictRow]" is not assignable to type "(__test: Any, /) -> RowMaker[TupleRow]"
        Function return type "RowMaker[DictRow]" is incompatible with type "RowMaker[TupleRow]"
          "RowMaker[DictRow]" is not assignable to "RowMaker[TupleRow]"
            Type parameter "Row@RowMaker" is covariant, but "DictRow" is not a subtype of "TupleRow"
      "FunctionType" is not assignable to "None" (reportArgumentType)
1 error, 0 warnings, 0 informations
```
